// @bevry/render/edition-es2022/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ma = exports.a = exports.mol = exports.ol = exports.mul = exports.ul = exports.li = exports.mcodeblock = exports.mcode = exports.code = exports.pre = exports.mblockquote = exports.blockquote = exports.mem = exports.em = exports.mstrong = exports.strong = exports.mh6 = exports.h6 = exports.mh5 = exports.h5 = exports.mh4 = exports.h4 = exports.mh3 = exports.h3 = exports.mh2 = exports.h2 = exports.mh1 = exports.h1 = exports.mh = exports.h = exports.mp = exports.p = exports.br = exports.t = exports.i = exports.tagEach = exports.tag = exports.wrapEach = exports.wrap = exports.trim = exports.join = exports.lines = exports.escapeAttribute = void 0;
/** Escape characters that can interfere with HTML and Markdown attributes */
function escapeAttribute(str) {
    return str
        .replaceAll('&', '&amp;') // must be done first
        .replaceAll('<', '&lt;')
        .replaceAll('>', '&gt;')
        .replaceAll('"', '&quot;') // double quote
        .replaceAll('[', '&#91;')
        .replaceAll(']', '&#93;')
        .replaceAll('(', '&#40;')
        .replaceAll(')', '&#41;')
        .replaceAll('`', '&#96;');
}
exports.escapeAttribute = escapeAttribute;
/** Combine lines that aren't empty */
function lines(items) {
    return typeof items === 'string' ? items : items.filter(Boolean).join('\n');
}
exports.lines = lines;
/** Join all lines */
function join(items, join = '\n') {
    return typeof items === 'string' ? items : items.join(join);
}
exports.join = join;
/** Trim whitespace from the start and end */
function trim(str) {
    return str.replace(/^\s+|\s+$/g, '');
}
exports.trim = trim;
/** Wrap the item */
function wrap(prefix = '', suffix = '', item) {
    if (item)
        return `${prefix}${item}${suffix}`;
    return '';
}
exports.wrap = wrap;
/** WRap each item */
function wrapEach(prefix = '', suffix = '', items) {
    if (typeof items === 'string')
        return [wrap(prefix, suffix, items)];
    return items.filter(Boolean).map((item) => wrap(prefix, suffix, item));
}
exports.wrapEach = wrapEach;
/** Wrap the item in a tag */
function tag(name = '', item) {
    return wrap(`<${name}>`, `</${name}>`, item);
}
exports.tag = tag;
/** Wrap each item in a tag */
function tagEach(name = '', items) {
    return wrapEach(`<${name}>`, `</${name}>`, items);
}
exports.tagEach = tagEach;
/** Render items conditionally */
function i(condition, items) {
    return condition ? lines(items()) : '';
}
exports.i = i;
/** Render text */
function t(items) {
    return typeof items === 'string' ? items : items.filter(Boolean).join(' ');
}
exports.t = t;
/** Render line break */
function br() {
    return '<br/>';
}
exports.br = br;
/** Render HTML paragraph */
function p(items) {
    return tag('p', lines(items));
}
exports.p = p;
/** Render Markdown paragraph */
function mp(items) {
    return lines(items) + '\n';
}
exports.mp = mp;
/** Render HTML heading */
function h(level, item) {
    return tag(`h${level}`, item);
}
exports.h = h;
/** Render Markdown heading */
function mh(level, item) {
    return `${'#'.repeat(level)} ${item}\n`;
}
exports.mh = mh;
/** Render HTML h1 */
function h1(item) {
    return h(1, item);
}
exports.h1 = h1;
/** Render Markdown h1 */
function mh1(item) {
    return mh(1, item);
}
exports.mh1 = mh1;
/** Render HTML h2 */
function h2(item) {
    return h(2, item);
}
exports.h2 = h2;
/** Render Markdown h2 */
function mh2(item) {
    return mh(2, item);
}
exports.mh2 = mh2;
/** Render HTML h3 */
function h3(item) {
    return h(3, item);
}
exports.h3 = h3;
/** Render Markdown h3 */
function mh3(item) {
    return mh(3, item);
}
exports.mh3 = mh3;
/** Render HTML h4 */
function h4(item) {
    return h(4, item);
}
exports.h4 = h4;
/** Render Markdown h4 */
function mh4(item) {
    return mh(4, item);
}
exports.mh4 = mh4;
/** Render HTML h5 */
function h5(item) {
    return h(5, item);
}
exports.h5 = h5;
/** Render Markdown h5 */
function mh5(item) {
    return mh(5, item);
}
exports.mh5 = mh5;
/** Render HTML h6 */
function h6(item) {
    return h(6, item);
}
exports.h6 = h6;
/** Render Markdown h6 */
function mh6(item) {
    return mh(6, item);
}
exports.mh6 = mh6;
/** Render HTML strong */
function strong(item) {
    return tag('strong', item);
}
exports.strong = strong;
/** Render Markdown strong */
function mstrong(item) {
    return `**${item}**`;
}
exports.mstrong = mstrong;
/** Render HTML em */
function em(item) {
    return tag('em', item);
}
exports.em = em;
/** Render Markdown em */
function mem(item) {
    return `_${item}_`;
}
exports.mem = mem;
/** Render HTML blockquote */
function blockquote(items) {
    return tag('blockquote', lines(items));
}
exports.blockquote = blockquote;
/** Render Markdown blockquote */
function mblockquote(items) {
    return wrapEach('> ', '', items).join('\n');
}
exports.mblockquote = mblockquote;
/** Render HTML pre */
function pre(items) {
    const content = join(items);
    if (!content)
        return '';
    return tag('pre', `\n${content}\n`);
}
exports.pre = pre;
/** Render HTML inline code */
function code(item) {
    return tag('code', item);
}
exports.code = code;
/** Render Markdown inline code */
function mcode(item) {
    return `\`${item}\``;
}
exports.mcode = mcode;
/** Render Markdown code block */
function mcodeblock(format, items) {
    const backticks = '```';
    const header = format ? `${backticks} ${format}` : backticks;
    return wrap(header + '\n', '\n' + backticks, join(items));
}
exports.mcodeblock = mcodeblock;
/** Render HTML list element */
function li(items) {
    return tagEach('li', items).join('\n');
}
exports.li = li;
/** Render HTML unordered list */
function ul(items) {
    return tag('ul', li(items));
}
exports.ul = ul;
/** Render Markdown unordered list */
function mul(items) {
    const content = wrapEach('-   ', '', items).join('\n');
    if (content)
        return content + '\n';
    return '';
}
exports.mul = mul;
/** Render HTML ordered list */
function ol(items) {
    return tag('ol', li(items));
}
exports.ol = ol;
/** Render Markdown ordered list */
function mol(items) {
    const content = wrapEach('1.  ', '', items).join('\n');
    if (content)
        return content + '\n';
    return '';
}
exports.mol = mol;
/** Render HTML link */
function a({ url, inner, title }) {
    if (!url || !inner) {
        throw new Error('Links must have both url and inner properties');
    }
    if (title) {
        return `<a href="${url}" title="${escapeAttribute(title)}">${inner}</a>`;
    }
    else {
        return `<a href="${url}">${inner}</a>`;
    }
}
exports.a = a;
/** Render Markdown link */
function ma({ url, inner, title }) {
    if (!url || !inner) {
        throw new Error('Links must have both url and inner properties');
    }
    if (title)
        return `[${inner}](${url} "${escapeAttribute(title)}")`; // @todo select ' or " to conform with prettier
    return `[${inner}](${url})`;
}
exports.ma = ma;
